### ISSUE TITLE
Causality in Temporal Convolutional Network (TCN)

### ISSUE THREAD

**Issue by StatMixedML** [unit8co/darts#480]
## Problem Description
I was going through your code of the [Temporal Convolutional Network (TCN)](https://github.com/unit8co/darts/blob/master/darts/models/tcn_model.py). I am not sure I understand the part of the residual connection. Potentially there is a problem with causality.

### Causality in Residual Connection in TCN 
Causality, as described [here,](https://github.com/locuslab/TCN/issues/8) can be achieved by 

```python
x = F.pad(x, ((k-1)*d, 0))
out = self.conv(x)
```
which is what you implemented in the forward method [here,](https://github.com/unit8co/darts/blob/master/darts/models/tcn_model.py#L94) reproduced as follows

```python
 def forward(self, x):
        residual = x

        # first step
        left_padding = (self.dilation_base ** self.nr_blocks_below) * (self.kernel_size - 1)
        x = F.pad(x, (left_padding, 0))
        x = self.dropout_fn(F.relu(self.conv1(x)))

        # second step
        x = F.pad(x, (left_padding, 0))
        x = self.conv2(x)
        if self.nr_blocks_below < self.num_layers - 1:
            x = F.relu(x)
        x = self.dropout_fn(x)

        # add residual
        if self.conv1.in_channels != self.conv2.out_channels:
            residual = self.conv3(residual)
        x += residual

        return x
```
I am ok with the `x` at each input step, since here causality is guaranteed by `left_padding` at each step. However, I struggle with the residual part `x += residual` since `residual`, which is the original `x`, is not causal so that we might add future information to `x` which is being returned. 

Appreciate your answer.

**Comment by pennfranc** [comment]
Hi @StatMixedML,
The term causality in this case is used as defined in the [original TCN paper](https://arxiv.org/pdf/1803.01271.pdf) as 'there can be no leakage from the future into the past'. This should not be violated when adding a residual, because it makes x_new at time t dependent on x_old at time t, which is allowed. The padding of x is required for the 1D convolution so that we don't convolve, say, x_old at times t+1 and t to create x_new at time t. But since we don't perform a convolution on the residual we don't have to pad it. Let me know if this makes sense to you (maybe I'm missing something)!

**Comment by StatMixedML** [comment]
@pennfranc Thanks for your fast response!

I understand that we need to ensure causality during the model training, i.e., for the causal convolutions. 

What I am struggling with is the residual connection piece after `conv1-3`: while we ensure no information leakage at time `t` from time points later than `t` during `conv1-3` we _simply_ add_ all information from the initial x to the output of the last `conv`, i.e., `x += residual`. Doesn't this imply that we violate causality since no special attention is paid to whether or not `t+1` must not be taken into account at `t` at any later dilation step.
